### DISCUSSION TITLE
IPFS helpers does not work with Node JS

### DISCUSSION THREAD

**Discussion by nicolapertosa** [capsule-corp-ternoa/ternoa-js#111] [General]
Dear Ternoa developers,

I installed 1.3.0-alpha0 Ternoa-js version because I wanted to test IPFS helpers for collections and NFT's metadata creations.
Actually, I'm using some Ternoa features in the back-end running with NodeJs and during invocation of collectionIpfsUpload, I get the following error message : 
`ReferenceError: FormData is not defined` in ternoa-js/helpers/ifps.js
In fact, FormData is native to the browser but not known in Node js and because there is no third party library imported in the ipfs helper to manage Files creation, it failes.

Could you please consider my request by using npm form-data dependency to avoid this error when using ipfs helpers with node JS.
`https://www.npmjs.com/package/form-data`

Thank you in advance for your help.

**Comment by ipapandinas** [comment]
hi @nicolapertosa,

Thank you for pointing this wrong behaviour out.
An **Alpha** release is intended to be used for internal testing. This feedback will help us to stabilise this helper before the upcoming **Release Candidate** version. I will come back here to provide a better solution after the `1.3.0-rc0` release.

**Comment by ipapandinas** [comment]
I faced the same error after testing this helper from a NodeJs back-end. Your pull request https://github.com/capsule-corp-ternoa/ternoa-js/pull/105 was merged and included in the `1.3.0-rc0` release ([changelog](https://github.com/capsule-corp-ternoa/ternoa-js/releases/tag/v1.3.0-rc0)).
This version will be tested for a few weeks before going stable. Let us know if everything works better on your side now.

Thank you so much for your help to fix this issue, you are the first external contributor to Ternoa.js 🏆

**Reply by nicolapertosa** [reply]
Hi @ipapandinas,
It's an honour being the first one contributing to this project as an external.

So I tested this IPFS upload once again and I notice now that I'm not able to upload any files on IPFS anymore... Did you change something in the IPFS API (https://ipfs.ternoa.dev/api/v0/add) ?

I also noticed in the helper that an header can be passed with apiKey property.
I guess this is not required if we want to use the default IPFS Gateway, right ?
By introducing this change of the header in the API, is it possible it could have an impact on the upload itself ?

Thank you for your answer.

**Reply by ipapandinas** [reply]
The IPFS dev API remained unchanged, this is a release that only concerns the SDK.
I just tested the API using the [Ternoa Workshop](https://workshop.ternoa.dev/) project and it's working fine.

You are right, the `apiKey` is not mandatory for our IPFS dev API. This change was already implemented in the previous alpha release, it should not impact the upload itself since it is optional.

In the future we are planing to move to an official Alphanet/Mainnet gateway which will need a proper `apiKey`. Developers will be able to claim their keys using a dedicated application (TBD).

Let us know if you are still facing issues.

**Reply by nicolapertosa** [reply]
Hi @ipapandinas,

I tested several times without success. 
Since 2 weeks now, I always get an HTTP 500 from axios when I try to upload an image to IPFS using "https://ipfs.ternoa.dev/api/v0/add" url.
This occurs always when it is a brand new image which is not yet uploaded on IPFS. However, if the image was already uploaded on IPFS before, then I can get the hash.

Do you have the possiblity to see the logs on the Ternoa IPFS instance and check what could be the reason of that http 500 response ?

Also, this is the code I use and which was working from de beginning:
```
import * as fs from 'fs';
import { ipfsFileUpload } from 'ternoa-js/helpers/ipfs.js'

(async () => {
    const file = fs.createReadStream('test.jpg');
    const result = await ipfsFileUpload(file);
    console.log(result);
})().catch((e) => {
    console.log(e);
});
```

Could you please test this code as well using only NodeJs and see if it works for you?
Thank you in advance for your help.

**Reply by Victor-Salomon** [reply]
@nicolapertosa, I think the issue on your code snippet comes from the fact that the type of your **file** variable is **fs.ReadStream** while the `ipfsFileUpload()` function from the **result** variable is expecting a parameter of type **File**. You need to convert your fs.ReadStream to a `new File()`

**Reply by nicolapertosa** [reply]
Hi @Victor-Salomon,

Unfortunately, there is no equivalent for "new File()" function in NodeJs.
However, in the form-data documentation, streams are allowed : https://www.npmjs.com/package/form-data

```
var FormData = require('form-data');
var fs = require('fs');

var form = new FormData();
form.append('my_field', 'my value');
form.append('my_buffer', new Buffer(10));
form.append('my_file', fs.createReadStream('/foo/bar.jpg'));
```

And looking further in ipfs-core documentation, buffers are handled as well : https://dev.to/rtagliavia/how-to-add-images-to-ipfs-with-nodejs-and-ipfs-core-3heo

```
import * as IPFS from 'ipfs-core'
import fs from 'fs'

const imagesDir = './src/images'

const files = fs.readdirSync(imagesDir)

const ipfs = await IPFS.create()

for(let file of files) {
  const buffer = fs.readFileSync(`${imagesDir}/${file}`)
  const result = await ipfs.add(buffer)
  console.log(result)
}
```

But this was working before with streams, so I don't know why this is happening now.

**Reply by Victor-Salomon** [reply]
@nicolapertosa let me come back to you as soon as possible with a solution.

**Reply by Victor-Salomon** [reply]
@nicolapertosa it should be good now. 

As explained by Igor (ipapandinas), we will move to an Alphanet/Mainnet gateway which will need a proper apiKey that you will have to request yourself. We are now focus on this, but this is not completely ready yet so we (re)setup the node on https://ipfs.ternoa.dev/. Please let me know if its good. I just set a small express server and it worked with your code on my side.

**Reply by nicolapertosa** [reply]
Hi @Victor-Salomon,

Thank you very much. This is working like a charm since you reactivated the node.
Don't hesitate to update us when the new IPFS Gateway is ready so that we can switch to the new one.

One more point, I was looking into the IPFS uploads with Ternoa-js SDK and I can already predict that the file nft/utils.ts will not work in NodeJs as well because Blob and File objects don't exist in that environment : 
https://github.com/capsule-corp-ternoa/ternoa-js/blob/main/src/nft/utils.ts (line : 31/32 and 55/56).

Do you want me to open a pull request and propose my solution for this ?

**Reply by Victor-Salomon** [reply]
@nicolapertosa, great reading that IPFS is working for you. We will keep an update about it in the every github discussion that concern IPFS. I already looked at the first key generator mockup/mvp and it looks mint 💎. 

_About, the others helpers (nft/collection and marketplace upload), I've had the same reflexion after launching the express server the other day. We used to upload on IPFS from our frontend UI, and as you said, it's not possible to directly use any new Blob or new Flie instances in a Node environement. I need to think about it a bit. The helper should be adaptable in a Node environement, since the generic IPFS upload is working properly._

We really appreciate your consideration and commitment into our products 🔥. Before we start working on it, feel free to help us on this. You can either open a pull request (according to our contribution [guidelines](https://github.com/capsule-corp-ternoa/ternoa-js/blob/main/CONTRIBUTING.md)), or keep discussing with us here.
